project_setup/docker/README.md

```diff
@@ -0,0 +1,264 @@
+and using [official Docker images](https://hub.docker.com/explore/) for Node,
+Postgres, Nginx, etc. We're working on a sustainable extension of the
+cflinuxfs2 image which would incorporate buildpacks, but in the meantime we
+recommend using the Ubuntu build (where available) of the official images,
+such as `python:3.5`, `ruby:2.4`, etc. Avoid the `slim` and `alpine` images as
+they aren't likely to closely match the production environment.
+
+### Mounting the working directory
+
+Demanding the developers rebuild their Docker images after every modification
+won't fly and isn't necessary. Instead, we'll want to mount the source code
+into a shared volume, meaning edits on the host file system are visible within
+the container and vis versa. This will likely look something like:
+
+```yml
+# docker-compose.yml
+my-service:
+  volumes:
+  - $PWD:/usr/src/app
+  working_dir: /usr/src/app
+  ...
+```
+
+where `/usr/src/app` is the mount point within the container (different Docker
+images will assume the project's code is at different spots within the file
+system; be sure to verify).
+
+By doing this, you guarantee that the container sees your changes without
+needing to rebuild the Docker image (which can take a considerable amount of
+time).
+
+Does this mean that your Dockerfile shouldn't include the application's source
+code? It certainly means that whatever the container _had_ stored at the
+mount point is ignored during local development. That doesn't rule out all of
+the benefits of including a checkpoint of the application source within the
+docker image, however. See [Are Dockerfiles needed?](#are-dockerfiles-needed)
+for further discussion.
+
+### Matching cloud.gov
+
+We'd like our local environment to match production, so let's mirror our
+cloud.gov manifest files in docker-compose.yml. This way, we'll be accessing
+our configuration variables the same way (e.g. via cf-env) in our two
+environments.
+
+Adding fixed environment variables (e.g. `DJANGO_SETTINGS_MODULE`) is straight
+forward, just add an `environment` key to your service definition. We'll also
+want to include settings provided within cloud.gov like `PORT`,
+`DATABASE_URL`, and (potentially) `TMPDIR`. See the [full
+list](https://docs.run.pivotal.io/devguide/deploy-apps/environment-variable.html#app-system-env)
+from cloud foundry. We'll also want to duplicate the serialized JSON of the
+`VCAP_APPLICATION` and `VCAP_SERVICES` variables, which is pretty straight
+forward using a "folded" string literal (`>`):
+
+```yml
+my-service:
+  environment:
+    DATABASE_URL: postgres://postgres@my-database-service/postgres
+    PORT: 8000
+    SOMETHING: Else
+    TMPDIR: /tmp
+    VCAP_APPLICATION: >
+      {"uris": ["localhost", "0.0.0.0", "127.0.0.1"]}
+    VCAP_SERVICES: >
+      {"my-user-provided-service": [{"credentials": {"SOME_KEY": "VALUE"}}],
+       "my-elastic-service": [{"more": "settings"}]}
+```
```

Comment:
I think we satisfy the "I want this to run in not-cloud.gov" need by running within Docker (even with the cloud.gov-looking config). Supporting different sources for configuration requires maintaining different documentation and tests for each, which has certainly bit me in the past.

In any event, I think it makes sense to exercise the cloud.gov-config path within Docker to have parity with production. This doesn't prevent other config paths from existing.